internal/cli/cli.go

```diff
@@ -0,0 +1,88 @@
+package cli
+
+import (
+	"io"
+
+	"github.com/10gen/realm-cli/internal/config"
+	"github.com/10gen/realm-cli/internal/realm"
+	"github.com/10gen/realm-cli/internal/ui"
+
+	"github.com/spf13/cobra"
+)
+
+// Command is an executable CLI command
+type Command interface {
+	Handler(args []string, io cmdio) error
+}
+
+// CommandPreparer handles the command setup phase
+type CommandPreparer interface {
+	Setup(args []string, io cmdio) error
+}
+
+// CommandResponder handles the command feedback phase
+type CommandResponder interface {
+	Feedback(args []string, io cmdio)
+}
+
+type cmdio struct {
+	in  io.Reader
+	err io.Writer
+	out io.Writer
+}
+
+func newCmdio(cmd *cobra.Command) cmdio {
+	return cmdio{
+		out: cmd.OutOrStdout(),
+		err: cmd.ErrOrStderr(),
+		in:  cmd.InOrStdin(),
+	}
+}
+
+func newCommand(use, short string, cmd Command) *cobra.Command {
+	var preRun cmdFnE
+	if c, ok := cmd.(CommandPreparer); ok {
+		preRun = c.Setup
+	}
+
+	var postRun cmdFn
+	if c, ok := cmd.(CommandResponder); ok {
+		postRun = c.Feedback
+	}
+
+	return &cobra.Command{
```

Comment:
[mega-nit] would just create the top level cobra.Command then add to it so we can avoid nil checks further down:
```	
c := cobra.Command{
		Use:     use,
		Short:   short,
		RunE:    cobraFnE(cmd.Handler),
}

	if cp, ok := cmd.(CommandPreparer); ok {
		c.PreRunE: cobraFnE(cp.Setup),
	}

	if cr, ok := cmd.(CommandResponder); ok {
		c.PostRun = cr.Feedback
	}
return &c
```